src/mgr/Action.cc

```diff
@@ -111,12 +111,35 @@ Mgr::Action::fillEntry(StoreEntry* entry, bool writeHttpHeader)
 
         entry->replaceHttpReply(rep);
     }
-
+    if (!aggregatable())
+        openKidSection(entry, is_yaml());
     dump(entry);
+    if (!aggregatable() && atomic())
+        closeKidSection(entry, is_yaml());
```

Comment:
Without the opening condition, this code should also be removed (and restored elsewhere!) AFAICT.

```suggestion
```

Reply:
That elsewhere is FunAction, which means that the framing would only be done for functions that register a callback function instead of the newer API. My understanding is that the idea is to eventually remove the callback-based API, is it not? @yadij had put in the work here IIRC, any input?

Reply:
> That elsewhere is FunAction, which means that the framing would only be done for functions that register a callback function instead of the newer API. 

Not exactly: FunAction (at least) will probably continue to frame kid-produced kid:ID responses. Other/aggregating actions will (eventually) start framing kids:IDs responses because only those actions know which kids they are polling to aggregate the reported info.

> My understanding is that the idea is to eventually remove the callback-based API, is it not?

Yes, at least for cases where there is some associated state that should be maintained by an Action-derived class.

Reply:
> Not exactly: FunAction (at least) will probably continue to frame kid-produced kid:ID responses. Other/aggregating actions will (eventually) start framing kids:IDs responses because only those actions know which kids they are polling to aggregate the reported info.

This I don't get. FunAction means (to me) FunctionAction which means it's job is to bridge between the new and the old API. Framing is outside its scope; it used to do it but there is no reason why this shouldn't be done one level up, so to be also useable for Actions that want to use the new API but not aggregate.
By going the direction you're suggesting, each Action using the new API needs to either aggregate or duplicate the framing. This feels to me like binding together concepts that do not need being bound

Reply:
> > FunAction (at least) will probably continue to frame kid-produced kid:ID responses. Other/aggregating actions will (eventually) start framing kids:IDs responses because only those actions know which kids they are polling to aggregate the reported info.

> This I don't get. FunAction means (to me) FunctionAction 

Correct.

> which means its job is to bridge between the new and the old API.

That assertion is not wrong, but it is _also_ possible to view FunAction as a class that provides a convenient way to implement certain simple cache manager actions. For example, if ten actions can be implemented well as ten basic stand-alone functions, then there is no need to define ten new classes, one for each of those ten implementations. I used this broader view of FunAction in my response. Needless to say, if we discover that there are virtually no such actions, then FunAction class will be gone when its last user is converted from a stand-alone function to a class.

> Framing is outside its scope; ... there is no reason why this shouldn't be done one level up

N.B. The current PR code does not move _all_ framing decisions "one level up", to Action -- there are still two more places where framing is currently done. I do not yet know whether doing all framing in Action is both feasible and a good idea.

For me, the correctness of the "scope" and "no reason" assertions quoted above remains to be seen: I am not yet ready to say which class or classes will be responsible for framing. We need to clean up and test this PR a bit further before I can confidently agree with those assertions. In general: If the base Action class can do some framing task well, then obviously it should be responsible for that task. Derived classes (possibly but not necessarily including FunAction) will continue to carry the remaining framing responsibilities (if any).

> so to be also useable for Actions that want to use the new API but not aggregate.

There is a third option here: Perhaps such Actions will have a common parent that is not Action. Again, for me, this design aspect remains to be seen.

> By going the direction you're suggesting, each Action using the new API needs to either aggregate or duplicate the framing.

If we do our job well, there will be no duplication. Please do not interpret my current change requests as an attempt to bring this PR (or even a portion of it) to its final state. They are a (failing) attempt to create a clean starting point and advance cautiously, in several iterations. The only feasible alternative I see in this context is for me to (slowly) define and implement this PR goals (going through these iterations without bothering you) and then post the final changes for your review. If you prefer that model, please let me know.

The affected PR code does not look good because it leaves a YAML document open for actions that are both not aggregatable() and not atomic(). It would be trivial to forget to close that document. The same problem exists in official code, but it is localized to FunAction where we may decide to ignore it. We should avoid making things worse, but if you insist on enlarging the scope/impact of this existing problematic code now, then I will come back to this problem later, after other PR problems have been resolved. Your call.

Reply:
> N.B. The current PR code does not move _all_ framing decisions "one level up", to Action -- there are still two more places where framing is currently done. I do not yet know whether doing all framing in Action is both feasible and a good idea.

Moving the framing fully to Action feels to me way more complicated than the benefit. The cleanest way I can think of to do it is to add and use a hook on storeEntry::completed() calling back into the Action. This is to cover the (so far only) non-atomic actions, which share most code.

There are three calls to closeKidAction() in the current code:
- one in Action which is where I'd like it to stay
- one in InfoAction. This is an artifact of the fact that InfoAction has a per-kid section, for which it is to provide its own framing. The correct solution for me is to remove that per-kid section to its own per-kid report, as we already discussed
- One in the objects-related code. This is necessary because it's a non-atomic action. Non-atomic actions are quite complicated, needing to have state, and continuations. They may end up having a million problems, but forgetting to close a kid section doesn't seem to be one to be concerned. And that is even assuming that we won't have a cleaner way to manage it using the stream-based API that Amos is wokrking on, or that at that point passing a functor to StoreEntry::completed() to close the kid will be a worthwhile endeavour

If we move the framing to FunAction, then every Action which is registered with the newer API will need to do its own framing, even when not atomic; or is there a way you envision to deal with this that I have missed?